arch-vega,gpu-compute,mem-ruby: SQC Invalidation Support

v-ramadas opened the pull request:
This PR adds support for SQC (GPU I-cache) invalidation to the GPU model. It does this by updating the GPU-VIPER-SQC protocol to support flushes, the sequencer model to send out invalidates and the gpu compute model to send invalidates and handle responses. It also adds support for S_ICACHE_INV, a VEGA ISA instruction that invalidates the entire GPU I-cache. Additionally, the PR modifies the kernel start behavior to invalidate the I-cache too. It previously invalidated only the L1 D-cache.

v-ramadas committed:
mem-ruby: Add SQC cache invalidation support  to GPU VIPER

This commit adds support for cache invalidation in GPU VIPER protocol's
SQC cache. To support this, the commit also adds L1 cache invalidation
framework in the Sequencer such that the Sequencer sends out an
invalidation request for each line in the cache and declares completion
once all lines are evicted.

Change-Id: I2f52eacabb2412b16f467f994e985c378230f841

v-ramadas committed:
gpu-compute: Add support for injecting scalar memory barrier

This commit adds support for injecting a scalar memory barrier in the
GPU. The barrier will primarily be used to invalidate the entire SQC
cache. The commit also invalidates all buffers and decrements related
counters upon completion of the invalidation request

Change-Id: Ib8e270bbeb8229a4470d606c96876ba5c87335bf

v-ramadas committed:
gpu-compute: Add Icache invalidation at kernel start

Previously, the data caches were invalidated at the start of each
kernel. This commit adds support for invalidating instruction cache at
kernel launch time

Change-Id: I32e50f63fa1442c2514d4dd8f9d7689759f503d3

mattsinc commented on src/arch/amdgpu/vega/insts/instructions.cc:
```diff
@@ -4818,6 +4818,10 @@ namespace VegaISA
+        setFlag(MemBarrier);
+        setFlag(GPUStaticInst::MemSync);
+        setFlag(MemSync);
+
```
extra line of space, remove

v-ramadas replied:
Done

mattsinc commented on src/gpu-compute/fetch_unit.cc:
```diff
@@ -388,6 +388,29 @@ FetchUnit::FetchBufDesc::flushBuf()
+    DPRINTF(GPUFetch, "WF[%d][%d]: Id%d Fetch dropped, flushing fetch "
+            "buffer\n", wavefront->simdId, wavefront->wfSlotId,
+            wavefront->wfDynId);
+
```
extra line of space, remove

mattsinc commented on src/gpu-compute/fetch_unit.cc:
```diff
@@ -471,6 +494,13 @@ FetchUnit::FetchBufDesc::reserveBuf(Addr vaddr)
 void
 FetchUnit::FetchBufDesc::fetchDone(Addr vaddr)
 {
+    if (vaddr == 0) {
```
you should probably add a comment here explaining what vaddr == 0 represents.

A bigger debate is if we can really guarantee vaddr == 0 only for s_icache_inv, but for now probably ok.

v-ramadas replied:
Added comment

mattsinc commented on src/gpu-compute/scalar_memory_pipeline.cc:
```diff
@@ -160,4 +160,55 @@ ScalarMemPipeline::issueRequest(GPUDynInstPtr gpuDynInst)
+
+    PacketPtr pkt = nullptr;
+
+    if (kernelMemSync) {
```
add comment here explaining if/else

v-ramadas replied:
Done

mattsinc commented on src/gpu-compute/scalar_memory_pipeline.cc:
```diff
@@ -160,4 +160,55 @@ ScalarMemPipeline::issueRequest(GPUDynInstPtr gpuDynInst)
+        req->requestorId(computeUnit.requestorId());
+    }
+
+    req->setPaddr(0);
```
add comment explaining this

v-ramadas replied:
Done

mattsinc commented on src/gpu-compute/scalar_memory_pipeline.cc:
```diff
@@ -160,4 +160,55 @@ ScalarMemPipeline::issueRequest(GPUDynInstPtr gpuDynInst)
+                new ComputeUnit::SQCPort::MemReqEvent
+                (computeUnit.sqcPort, pkt);
+
+        computeUnit.schedule(
```
probably the schedule can be after the if-else, since it's the same in both?  And maybe the sqc_event creation too?

v-ramadas replied:
Done

mattsinc commented on src/gpu-compute/shader.cc:
```diff
@@ -221,6 +221,11 @@ Shader::prepareInvalidate(HSAQueueEntry *task) {
         cuList[i_cu]->doInvalidate(req, task->dispatchId());
 
+
+        if ((i_cu % 4) == 0) {
```
Do we have a variable in the Python for this?  4 is a magic number because of how many CUs share an SQC.  To make this code more resilient we should really add/use such a variable to avoid problems later.

v-ramadas replied:
We have one in the top level config script but did not in the Shader/Compute Unit models. I added one and use it in this logic instead. It gets its value from the top level config script. I wanted to do this in my initial commit but completely forgot about it. Thanks for pointing it out.

mattsinc replied:
Probably we'd need to update that script too to reflect this change, but it shouldn't be part of this patch

mattsinc commented on src/gpu-compute/shader.cc:
```diff
@@ -221,6 +221,11 @@ Shader::prepareInvalidate(HSAQueueEntry *task) {
         cuList[i_cu]->doInvalidate(req, task->dispatchId());
 
+
+        if ((i_cu % 4) == 0) {
```
also add comment here explaining logic

v-ramadas replied:
Done

mattsinc commented on src/mem/ruby/protocol/GPU_VIPER-SQC.sm:
```diff
@@ -313,6 +320,11 @@ machine(MachineType:SQC, "GPU SQC (L1 I Cache)")
+  action(inv_invDone, "inv", desc="local inv done") {
+    sequencer.invL1Callback();
+  }
+
```
extra line of space, remove

v-ramadas replied:
Done

mattsinc commented on src/mem/ruby/system/Sequencer.cc:
```diff
@@ -348,6 +350,10 @@ Sequencer::insertRequest(PacketPtr pkt, RubyRequestType primary_type,
     }
 
+    if (pkt->cmd == MemCmd::MemSyncReq) {
+        return RequestStatus_Aliased;
```
What is this for?

v-ramadas replied:
Added a comment explaining it

mattsinc reviewed (changes requested):
In general looks good, but some questions/comments/changes.  Nice job!

v-ramadas committed:
arch-vega, gpu-compute, mem-ruby: Remove extra empty lines

Change-Id: I18770ec7e38c4a992a0ae6de95b0be49ab4426c2

v-ramadas committed:
configs, gpu-compute: Add parameter in shader for CUs per SQC

Change-Id: If0ae0db1b6ccc08a92f169a271b137f69f410f7b

v-ramadas committed:
gpu-compute, mem-ruby: Add comments and reformat code

Change-Id: Id2b3886dce347fdcfcad22009a42b92febc00a6c

v-ramadas commented:
@mattsinc, thanks for reviewing! I have made the updates you suggested and added comments accordingly.

mattsinc commented on src/mem/ruby/protocol/MOESI_AMD_Base-CorePair.sm:
```diff
@@ -575,9 +581,12 @@ machine(MachineType:CorePair, "CP-like Core Coherence")
                     trigger(Event:C1_Load_L1miss, in_msg.LineAddress,
                             cache_entry, tbe);
-                  } else {
+                  } else if ((in_msg.Type == RubyRequestType:ST) || (in_msg.Type == RubyRequestType:ATOMIC)) {
```
What does this have to do with the SQC commit?

v-ramadas replied:
I incorrectly added this file. These changes were in my stash and I applied it wrongly added it when making review changes. Removed this from the commit now.

These changes are from when I added Flush support for L1,L2. Since GPUFS uses KVM instead of the CorePair, I didn't get to test this. Haven't made a commit for these changes because of this reason

mattsinc commented on src/mem/ruby/protocol/MOESI_AMD_Base-CorePair.sm:
```diff
@@ -613,9 +625,12 @@ machine(MachineType:CorePair, "CP-like Core Coherence")
                     trigger(Event:C0_Load_L1miss, in_msg.LineAddress,
                         cache_entry, tbe);
-                  } else {
+                  } else if ((in_msg.Type == RubyRequestType:ST) || (in_msg.Type == RubyRequestType:ATOMIC)) {
```
same here

abmerop commented on src/gpu-compute/compute_unit.cc:
```diff
@@ -1042,6 +1069,28 @@ ComputeUnit::SQCPort::recvReqRetry()
+    SenderState *sender_state = safe_cast<SenderState*>(pkt->senderState);
+    [[maybe_unused]] ComputeUnit *compute_unit = sqcPort.computeUnit;
+
+    if (pkt->req->systemReq()) {
+        assert(compute_unit->shader->systemHub);
+        SystemHubEvent *resp_event = new SystemHubEvent(pkt, &sqcPort);
+        compute_unit->shader->systemHub->sendRequest(pkt, resp_event);
```
I don't think this path is possible.  If it is, since the SystemHubEvent process() does nothing we will have requests silently dropped.  I would instead make this an assert that it is not a system request.  If we see the assert in the future we'll know exactly where to go back and fix it

v-ramadas replied:
Thanks for letting me know. I was not sure about this part earlier. I have now changed this to an assert and the code still works

abmerop commented on src/gpu-compute/compute_unit.hh:
```diff
@@ -674,6 +675,41 @@ class ComputeUnit : public ClockedObject
+            const char *description() const;
+        };
+
+        class SystemHubEvent : public Event
+        {
+          SQCPort *sqcPort;
+          PacketPtr reqPkt;
+
+          public:
+            SystemHubEvent(PacketPtr pkt, SQCPort *_sqcPort)
+                : sqcPort(_sqcPort), reqPkt(pkt)
+            {
+                setFlags(Event::AutoDelete);
+            }
+
+            void
+            process()
+            {
+            }
+        };
+
```
Probably not needed due to above comment

v-ramadas replied:
Removed

abmerop commented on src/gpu-compute/fetch_unit.cc:
```diff
@@ -388,6 +388,28 @@ FetchUnit::FetchBufDesc::flushBuf()
             wavefront->wfDynId);
 }
 
+void
+FetchUnit::FetchBufDesc::invBuf()
+{
+    restartFromBranch = false;
+    /**
+     * free list may have some entries
+     * so we clear it here to avoid duplicates
+     */
+    freeList.clear();
+    bufferedPCs.clear();
+    reservedPCs.clear();
+    readPtr = bufStart;
+
+    for (int i = 0; i < fetchDepth; ++i) {
+        freeList.push_back(bufStart + i * cacheLineSize);
+    }
+
+    DPRINTF(GPUFetch, "WF[%d][%d]: Id%d Fetch dropped, flushing fetch "
+            "buffer\n", wavefront->simdId, wavefront->wfSlotId,
+            wavefront->wfDynId);
+}
+
```
This seems to be exactly the same as flushBuf(), is there any reason not to use that?  I realize the terms have different meanings, but for a read-only buffer they are functionally the same

abmerop commented on src/gpu-compute/fetch_unit.cc:
```diff
@@ -471,6 +493,18 @@ FetchUnit::FetchBufDesc::reserveBuf(Addr vaddr)
+    // associated with this instruction, decLGKMInstsIssued() is not
+    // executed. Do this here to decrement the counter and invalidate
+    // all buffers
+    if (vaddr == 0) {
```
This method seems to be called in one place only which has access to the packet.  I think it would be safer to change the method to pass the packet, then here it could check if it is a MemSyncResp (or whatever SQC invalidate response is)

v-ramadas replied:
I updated the function to use the packet command instead. Thanks for the suggestion!

mattsinc replied:
good call, @abmerop !

v-ramadas committed:
gpu-compute: Remove unused and redundant functions

In ComputeUnit, a previous commit added a  SystemHubEvent event class to
the SQCPort. This was found to be unnecessary during the review process
and is removed in this commit. Similarly, invBuf() which was added in
FetchUnit as part of an earlier commit was found to be redundant. This
commit removes it

Change-Id: I6ee8d344d29e7bfade49fb9549654b71e3c4b96f

abmerop commented on src/arch/amdgpu/vega/insts/instructions.cc:
Forgot this file no longer exists.  It would now go in the sopp.cc file

mattsinc replied:
good point @abmerop , @v-ramadas you must not have updated to the head of develop before making your changes?

v-ramadas replied:
I was at the head of develop when I first made the branch. This was before the commit to break up instructions.cc into multiple files was merged though. I will update the PR with this change now

v-ramadas committed:
arch-vega: Add support for S_ICACHE_INV instruction

Previously, the S_ICACHE_INV instruction was unimplemented and
simulation panicked if it was encountered. This commit adds support for
executing the instruction by injecting a memory barrier in the scalar
pipeline and invalidating the ICACHE (or SQC)

Change-Id: I0fbd4e53f630a267971a23cea6f17d4fef403d15

v-ramadas commented:
@mattsinc @abmerop, I rebased my branch to the top of develop and wrote the S_ICACHE_INV behavior in sopp.cc. I also removed the older commit that was causing the merge issue. I've pushed the changes to this PR after verifying my test still works.
